### DISCUSSION TITLE
Debug mode

### DISCUSSION THREAD

**Discussion by Geoffrey-Fish** [AntMeNet/AntMeClassic#57] [Q&A] [answered]
Hi, 
da der letzte offene Beitrag erst 2 Wochen her ist,bin ich erstaunt,dass dieses Projekt noch gesehen wird!
Ich wollte es nochmal probieren(das letzte mal war iwan 2015) und ich würde gern meine meisen live debuggen und sehen, was da so droppt an infos. Es gibt ja das Think modul, aber ich finde nirgends die Möglichkeit, eine Console o.ä. dazu zu schalten.
Help!
Merci,
Geoff

**Comment by patteki** [comment, marked as answer]
Guten Abend Geoffrey,
schön dass noch an AntMe! dran bist und dein Können erweitern möchtest.
Zum Debuggen gibt es prinzipiell Zwei Möglichkeiten:

### 1. Ingame Debug-Ansicht (Einsteiger)
Wie du schon bemerkt hast, das die Think (bzw. im deutschen die Denke) Methode wenn du diese im Code aufrufst und ihr einen Text in form eines String Wertes übergibts, kannst du dir diesen in der 3D-Ansicht als Text-(Denk-)blase anzeigen lassen.
Um diese zu aktivieren, musst du während die 3D-Ansicht läuft auf die Taste "D" drücken, diese öffnet die Debug-Ansicht jetzt kannst du auch die Blasenüber den Ameisen sehen. Außerdem siehst du dann auch den aktuellen Sichtradius der Ameise und auch ihr Aktuelles Ziel als Kreise und Linien.

### 2. VisualStudio Debugger (Fortgeschritten)
Vieleicht hast du schon einmal bemerkt, dass du im Code-Editor eine Zeile rot markieren kannst, in dem du ganz vorne in die Zeile auf den grauen Rand klickst (da wo der Rote kreis dann erscheint).
<img width="344" alt="image" src="https://github.com/AntMeNet/AntMeClassic/assets/13452239/b9d2c0e0-0ed9-46a1-bbcf-d433a2040b34">

nun ist diese Zeile als sogenannter "Breakpoint" gesetzt dort wird also das Program anhalten und auf dich warten, wenn es soweit abgelaufen ist.
Nun drückst du F5 oder im Menü Debuggen auf Debuggen starten.
Darauf hin öffnet sich AntMe! und du musst nur noch die Simulation starten.
Wenn nun die Simulation an einen Breakpoint kommt, wird sie automatisch pausieren und auf dich warten.
So sieht das dann in VisualStudio aus:

<img width="315" alt="image" src="https://github.com/AntMeNet/AntMeClassic/assets/13452239/5002999a-9451-453c-9bcc-27b3ae6073a9">

Nun ist das Programm angehalten und du kannst Variablen usw. erkunden und dir die Werte genau anschauen, in dem du z.b. mit der Maus über eine Variable fährst.
Dort öffnet sich dann ein kleines Fenster, in dem du alle Werte sehen kannst.
Wenn du nun z.b. Schritt für Schritt durch dein Programm gehen willst um zu sehen wie sich Variablen Werte verändern, kannst du mit F10 jeweils einen Schritt weiter gehen.
Mit F5 läuft dein Programm dann ganz normal weiter, bis zum nächsten Breakpoint.

Eine Warnung muss ich dir noch mit auf den Weg der 2. Variante geben. Du hast im Normalen Spiel nicht nur 1 Ameise und wenn z.b. gerade 10 Ameisen Müde werden, wirst du 10 mal an den Breakpoint hängen bleibst bevor der nächste Simulationsschritt passiert.

Ich hoffe ich konnte dir damit weiterhelfen und du kannst nun deine Ameisen noch besser analysieren und ihr Verhalten verbessern.

Schönen Abend noch und frohes Krabbeln
LG Patrick
